File: src/main/kotlin/ordertaking/common/Entity.kt
package com.kalrap.ordertaking.common

abstract class Entity<T: Any> {
    abstract val id: T

    override fun equals(other: Any?): Boolean {
       return this === other || (
               other != null &&
               this::class == other::class &&
               other is Entity<*> &&
               this.id == other.id
               )
    }

    override fun hashCode(): Int = id.hashCode()
}

File: ts/src/ordertaking/common/entity.ts
type RawId = string | number | bigint;

export abstract class Entity<ID extends RawId | ValueObject> implements Equatable {
    abstract readonly id: ID;
    protected abstract isSameClass<T extends Entity<ID>>(obj: unknown): obj is T;

    @bound
    equals(obj: unknown): boolean {
        if (!this.isSameClass(obj)) return false;
        const otherId = (obj as Entity<ID>).id;

        return this.id instanceof ValueObject
        ? this.id.equals(otherId)
            : this.id === otherId;
    }
}